runtime/drivers/bigquery/sql_store.go

```diff
@@ -43,41 +44,93 @@ func (c *Connection) QueryAsFiles(ctx context.Context, props map[string]any, opt
 		return nil, err
 	}
 
-	client, err := bigquery.NewClient(ctx, srcProps.ProjectID, opts...)
-	if err != nil {
-		if strings.Contains(err.Error(), "unable to detect projectID") {
-			return nil, fmt.Errorf("projectID not detected in credentials. Please set `project_id` in source yaml")
+	now := time.Now()
+
+	var client *bigquery.Client
+	var it *bigquery.RowIterator
+	bqClientFn := func(projectID string) (*bigquery.Client, error) {
+		client, err := bigquery.NewClient(ctx, projectID, opts...)
+		if err != nil {
+			if strings.Contains(err.Error(), "unable to detect projectID") {
+				return nil, fmt.Errorf("projectID not detected in credentials. Please set `project_id` in source yaml")
+			}
+			return nil, fmt.Errorf("failed to create bigquery client: %w", err)
 		}
-		return nil, fmt.Errorf("failed to create bigquery client: %w", err)
+		return client, nil
 	}
 
-	if err := client.EnableStorageReadClient(ctx, opts...); err != nil {
-		client.Close()
-		return nil, err
-	}
+	r := regexp.MustCompile("(?i)^\\s*SELECT\\s+\\*\\s+FROM\\s+(`?[a-zA-Z0-9_.-]+`?)\\s*$")
+	match := r.FindStringSubmatch(srcProps.SQL)
+	if match != nil {
+		// "SELECT * FROM `project_id.dataset.table`" statement so storage api might be used
+		// project_id, dataset, backticks are optional
+		fullTableName := match[1]
+		fullTableName = strings.Trim(fullTableName, "`")
+
+		var projectID, dataset, tableID string
+
+		parts := strings.Split(fullTableName, ".")
+		switch len(parts) {
+		case 1:
+			tableID = parts[0]
+		case 2:
+			dataset, tableID = parts[0], parts[1]
+		case 3:
+			projectID, dataset, tableID = parts[0], parts[1], parts[2]
+		default:
+			return nil, fmt.Errorf("invalid table format, `project_id.dataset.table` is expected")
+		}
 
-	now := time.Now()
-	q := client.Query(srcProps.SQL)
-	it, err := q.Read(ctx)
-	if err != nil && !strings.Contains(err.Error(), "Syntax error") {
-		// close the read storage API client
-		client.Close()
-		c.logger.Info("query failed, retrying without storage api", zap.Error(err))
-		// the query results are always cached in a temporary table that storage api can use
-		// there are some exceptions when results aren't cached
-		// so we also try without storage api
-		client, err = bigquery.NewClient(ctx, srcProps.ProjectID, opts...)
+		// Override projectID if a user-defined is set
+		if srcProps.ProjectID != "" && srcProps.ProjectID != bigquery.DetectProjectID {
```

Comment:
I think we can't rely on the `project_id` passed in query for all cases. User may be trying to read data from another project where they don't have access to compute resources.
The project should always be either detected from credentials or passed explicitly in `project_id` field.

Reply:
It should be pretty common to read data from public datasets.

Reply:
Updated the logic so that If source project id isn't passed via a query then a corresponding source property is used